### PR TITLE
Support for list extensions

### PR DESCRIPTION
**Pull request by codito**
Add following command line switches to `vstest.console.exe`:
* /listDiscoverers
* /listExecutors
* /listLoggers
* /listSettingsProviders

These will list the corresponding test platform extensions. They are added for compatibility with earlier version of `vstest.console.exe`.

### CODE REVIEW

File: src/vstest.console/Processors/ListExtensionsArgumentProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+            {
+                ConsoleOutput.Instance.WriteLine(extension.Value.GetType().FullName, OutputLevel.Information);
+                ConsoleOutput.Instance.WriteLine(string.Format(CultureInfo.CurrentCulture, CommandLineResources.AvailableExtensionsMetadataFormat, "SettingName", extension.Metadata.SettingsName), OutputLevel.Information);
+            }
+
+            return ArgumentProcessorResult.Success;
+        }
+
+        private class NullMessageLogger : IMessageLogger
+        {
```
Comment (chakradhar2) [inline src/vstest.console/Processors/ListExtensionsArgumentProcessor.cs]: Do we require NullMessageLogger here?

Comment (codito) [reply src/vstest.console/Processors/ListExtensionsArgumentProcessor.cs]: It is required by this argument processor.

Comment (codito) [reply src/vstest.console/Processors/ListExtensionsArgumentProcessor.cs]: @chakradhar2 thanks, I've removed the extraneous code.

File: src/vstest.console/Processors/ListExtensionsArgumentProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,236 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
```
Comment (singhsarab) [inline src/vstest.console/Processors/ListExtensionsArgumentProcessor.cs]: Are these getting deprecated ? If yes, should we let the user know that these options are getting deprecated via message.

**Review by singhsarab — APPROVED** [review]
With a suggestion.

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginManager.cs [hunk 3]
```diff
@@ -5,6 +5,7 @@ namespace Microsoft.VisualStudio.TestPlatform.Common.ExtensionFramework
 {
     using System;
     using System.Collections.Generic;
+    using System.Linq;
```
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginManager.cs]: nit: Removed unused using

File: test/Microsoft.TestPlatform.AcceptanceTests/ListExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    [TestClass]
+    public class ListExtensionsTests : AcceptanceTestBase
+    {
+        [CustomDataTestMethod]
+        [NETFullTargetFramework(inIsolation: false, inProcess: true)]
+        public void ListDiscoverersShouldShowInboxDiscoverers(RunnnerInfo runnerInfo)
+        {
+            AcceptanceTestBase.SetTestEnvironment(this.testEnvironment, runnerInfo);
```
Comment (smadala) [inline test/Microsoft.TestPlatform.AcceptanceTests/ListExtensionsTests.cs]: Is it okay to create one test for all arguments?

File: src/vstest.console/Processors/ListExtensionsArgumentProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+    internal class ListDiscoverersArgumentExecutor : IArgumentExecutor
+    {
+        public void Initialize(string argument)
+        {
+        }
+
+        public ArgumentProcessorResult Execute()
+        {
+            ConsoleOutput.Instance.WriteLine(CommandLineResources.AvailableDiscoverersHeaderMessage, OutputLevel.Information);
```
Comment (smadala) [inline src/vstest.console/Processors/ListExtensionsArgumentProcessor.cs]: How come `AvailableDiscoverersHeaderMessage` resources strings already exists? Are we having any more unused resources strings?

**Review by smadala — APPROVED** [review]

**Comment by smadala** [comment]
What is the behavior when running using `dotnet vstest`?

**Review by Faizan2304 — APPROVED** [review]

**Comment by Faizan2304** [comment]
@codito : I am assuming it will discover extensions from --TeatAdapterPath as well.

**Comment by smadala** [comment]
`vstest.console.exe /TestAdapterPath:"Path\to\core-adapters" /listDiscoverers`, will show the discoverer?

File: src/vstest.console/Processors/ListExtensionsArgumentProcessor.cs (hunk 5 above)
Comment (codito) [reply src/vstest.console/Processors/ListExtensionsArgumentProcessor.cs]: I think these were copied as-is from older version of test platform.

**Comment by codito** [comment]
@smadala @Faizan2304 no, this scenario is not supported. In tpv2, the runner process doesn't load the external test adapters; they are always known in the test host (which is specific to user provided runtime). This is different than tpv1 which only supported test adapters authored in .NET full framework.

**Pull request merged**